# Day 4: Secure Container

# Part 1

You arrive at the Venus fuel depot only to discover it's protected by a password. The Elves had written the password on a sticky note, but someone threw it out.

However, they do remember a few key facts about the password:

- It is a six-digit number.
- The value is within the range given in your puzzle input.
- Two adjacent digits are the same (like `22` in `1`**`22`**`345`).
- Going from left to right, the digits **never decrease**; they only     ever increase or stay the same (like `111123` or `135679`).

Other than the range rule, the following are true:

- `111111` meets these criteria (double `11`, never decreases).
- `2234`**`50`** does not meet these criteria (decreasing pair of          digits `50`).
- `123789` does not meet these criteria (no double).

**How many different passwords** within the range given in your puzzle input meet these criteria?

In [1]:
inputs = range(347312, 805916)

In [2]:
def adjacent_equals(password):
    twos = zip(str(password), str(password)[1:])
    twos = [first == second for first, second in twos]
    return any(twos)

In [3]:
def not_decreasing(password):
    twos = zip(str(password), str(password)[1:])
    twos = [first <= second for first, second in twos]
    return all(twos)

In [4]:
def is_valid(password):
    return adjacent_equals(password) and not_decreasing(password)

In [5]:
def num_valid(passwords):
    validity = [is_valid(password) for password in passwords]
    num_valid = sum(validity)
    return num_valid

### Test Cases

In [6]:
test_cases = [
    (111111, True),
    (223450, False),
    (123789, False)
]

for input_, output in test_cases:
    assert is_valid(input_) is output

In [7]:
num_valid(inputs)

594

Answer to Part 1: **594**

## Part 2

An Elf just remembered one more important detail: the two adjacent matching digits **are not part of a larger group of matching digits**.

Given this additional criterion, but still ignoring the range rule, the following are now true:

- `112233` meets these criteria because the digits never decrease and     all repeated digits are exactly two digits long.
- `123`**`444`** no longer meets the criteria (the repeated `44` is       part of a larger group of `444`).
- `111122` meets the criteria (even though `1` is repeated more than     twice, it still contains a double `22`).

**How many different passwords** within the range given in your puzzle input meet all of the criteria?

In [8]:
import itertools

In [9]:
def has_only_two(password):
    grouped = itertools.groupby(str(password))
    counts = (len(list(group)) for _, group in grouped)
    two_counts = (count == 2 for count in counts)
    return any(two_counts)

In [10]:
def is_valid(pw):
    return adjacent_equals(pw) and not_decreasing(pw) and has_only_two(pw)

In [11]:
def num_valid(passwords):
    validity = [is_valid(password) for password in passwords]
    num_valid = sum(validity)
    return num_valid

### Test Cases

In [12]:
test_cases = [
    (112233, True),
    (123444, False),
    (111122, True)
]

for input_, output in test_cases:
    assert is_valid(input_) is output

In [13]:
num_valid(inputs)

364

Answer to Part 2: **364**